x/pairing/keeper/scores/score.go

```diff
@@ -0,0 +1,213 @@
+		slots[i] = NewPairingSlot()
+		slots[i].Reqs = slotReqs
+	}
+
+	return slots
+}
+
+// group the slots
+func GroupSlots(slots []*PairingSlot) []*PairingSlot {
+	uniqueSlots := []*PairingSlot{}
+
+	if len(slots) == 0 {
+		panic("no pairing slots available")
+	}
+
+	uniqueSlots = append(uniqueSlots, slots[0])
+	for k := 1; k < len(slots); k++ {
+		isUnique := true
+
+		for i := range uniqueSlots {
+			if slots[k].Equal(uniqueSlots[i]) {
+				uniqueSlots[i].Count += 1
+				isUnique = false
+				break
+			}
+		}
+
+		if isUnique {
+			uniqueSlots = append(uniqueSlots, slots[k])
+		}
+	}
+
+	return uniqueSlots
+}
+
+// TODO: currently we'll use weight=1 for all reqs. In the future, we'll get it from policy
+func GetStrategy() ScoreStrategy {
+	return uniformStrategy
+}
+
+// CalcPairingScore calculates the final pairing score for a pairing slot (with strategy)
+// For efficiency purposes, we calculate the score on a diff slot which represents the diff reqs of the current slot
+// and the previous slot
+func CalcPairingScore(scores []*PairingScore, strategy ScoreStrategy, diffSlot *PairingSlot, minStake sdk.Int) error {
+	// calculate the score for each req for each provider
+	for _, req := range diffSlot.Reqs {
+		reqName := req.GetName()
+		weight := strategy[reqName] // not checking if key is found because it's verified in init()
+
+		for _, score := range scores {
+			newScoreComp := req.Score(*score.Provider)
+			if newScoreComp == 0 {
+				err := fmt.Errorf("score component is zero. score component name: %s, provider address: %s", reqName, score.Provider.Address)
+				panic(err)
+			}
+			newScoreComp = uint64(math.Pow(float64(newScoreComp), float64(weight)))
+
+			// divide by previous score component (if exists) and multiply by new score
+			prevReqScoreComp, ok := score.ScoreComponents[reqName]
+			if ok {
+				score.Score /= prevReqScoreComp
+			}
+			score.Score *= newScoreComp
+
+			// update the score component map
+			score.ScoreComponents[reqName] = newScoreComp
+		}
+	}
+
+	return nil
+}
+
+// PrepareHashData prepares the hash needed in the pseudo-random choice of providers
+func PrepareHashData(projectIndex string, chainID string, epochHash []byte) []byte {
+	hashData := []byte{}
+
+	// add the session start block hash to the function to make it as unpredictable as we can
+	hashData = append(hashData, epochHash...)
+	hashData = append(hashData, chainID...)      // to make this pairing unique per chainID
+	hashData = append(hashData, projectIndex...) // to make this pairing unique per consumer
+
+	return hashData
+}
+
+// PickProviders pick a <group-count> providers set with a pseudo-random weighted choice (using the providers' score list and hashData)
+func PickProviders(ctx sdk.Context, scores []*PairingScore, groupCount uint64, hashData []byte, chosenProvidersIdx *map[int]bool) (returnedProviders []epochstoragetypes.StakeEntry) {
+	if len(scores) == 0 {
+		return returnedProviders
+	}
+
+	scoreSum := sdk.NewUint(0)
```

Comment:
it would be more efficient to sum the scores as uin64, and then convert with NewUint().
you can also add FindSum() generic (in common/types/utils.go) for that.

Reply:
actually, why even use `sdk.Int` instead of `uint64` all through here?

Reply:
Go's native mod operation is `%`

Reply:
done. see https://github.com/lavanet/lava/commit/5bdce0ff0c83e5d6f133414306b05605d746e1be